internal/devpkg/narinfo_cache.go

```diff
@@ -103,39 +72,34 @@ func FillNarInfoCache(ctx context.Context, packages ...*Package) error {
 
 	group, _ := errgroup.WithContext(ctx)
 	for _, p := range eligiblePackages {
-		// If the package's NarInfo status is already known, skip it
-		isNarInfoInCache.lock.RLock()
-		_, ok := isNarInfoInCache.status[p.Raw]
-		isNarInfoInCache.lock.RUnlock()
-		if ok {
-			continue
-		}
 		pkg := p // copy the loop variable since its used in a closure below
 		group.Go(func() error {
-			err := pkg.fillNarInfoCache()
-			if err != nil {
-				// default to false if there was an error, so we don't re-try
-				isNarInfoInCache.lock.Lock()
-				isNarInfoInCache.status[pkg.Raw] = false
-				isNarInfoInCache.lock.Unlock()
-			}
+			_, err := pkg.fillNarInfoCacheIfNeeded()
 			return err
 		})
 	}
 	return group.Wait()
 }
 
-// fillNarInfoCache fills the cache value for the narinfo of this package,
+// fillNarInfoCacheIfNeeded fills the cache value for the narinfo of this package,
 // assuming it is eligible for the binary cache. Callers are responsible
 // for checking isEligibleForBinaryCache before calling this function.
 //
 // NOTE: this must be concurrency safe.
-func (p *Package) fillNarInfoCache() error {
+func (p *Package) fillNarInfoCacheIfNeeded() (bool, error) {
+	if status, alreadySet := isNarInfoInCache.Load(p.Raw); alreadySet {
+		return status.(bool), nil
+	}
+	defer func() {
+		// If there's an error and nothing got cached, just store false.
+		_, _ = isNarInfoInCache.LoadOrStore(p.Raw, false)
+	}()
```

Comment:
@gcurtis @savil using defer here is functionally identical to what was previously handled in line 119 except now it is handled 100% of the time. 

This in conjunction with `LoadOrStore` in line 125 guarantee that the first result is preserved forever (so a race condition cannot replace an existing value). This fixes an existing race condition where two calls to `fillNarInfoCacheIfNeeded` could override each other if they had different results.

This fix also removes the race condition where an early `false` causes another goroutine to return `false` early.

It partially* fixes multiple goroutines trying to fetch the same result at the same time. In that case only the first result is saved and other goroutines return the same result. Previously, each call to `fillNarInfoCacheIfNeeded` could return a different result.

* I say "partially" because ideally we don't make 2 http requests, but I think that's not a big deal so unless the fix is easy, I would avoid complicating the code.

Reply:
> This fix also removes the race condition where an early false causes another goroutine to return false early.

I don't quite understand this scenario. Care to elaborate?

> I say "partially" because ideally we don't make 2 http requests, but I think that's not a big deal so unless the fix is easy, I would avoid complicating the code.

Did you try @gcurtis 's suggestion of using `sync.Once`s? I think that would address this.